0.12-rc1 crashes on Terraform apply when specifying input plan

### Terraform Version
```
Terraform v0.12.0-rc1
+ provider.aws v2.9.0
```

### Crash Output
```
  meta.count-boundary (EachMode fixup) - *terraform.Npanic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1921737]

goroutine 699 [running]:
github.com/hashicorp/terraform/terraform.(*NodeApplyableResourceInstance).EvalTree(0xc000c02d50, 0x276a220, 0xc000c02d50)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_apply_instance.go:111 +0x2a7
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x276a220, 0xc000c02d50, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:78 +0xd30
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc0003e4180, 0x276a220, 0xc000c02d50, 0xc000d924c0)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x353
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xa9b
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1921737]

goroutine 699 [running]:
github.com/hashicorp/terraform/terraform.(*NodeApplyableResourceInstance).EvalTree(0xc000c02d50, 0x276a220, 0xc000c02d50)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_apply_instance.go:111 +0x2a7
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x276a220, 0xc000c02d50, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:78 +0xd30
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc0003e4180, 0x276a220, 0xc000c02d50, 0xc000d924c0)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x353
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xa9b
```

### Expected Behavior
For the plan to be applied.

### Actual Behavior
The Terraform CLI crashes

### Steps to Reproduce
`terraform plan -out=terraform.plan`
`terraform apply terraform.plan`

--- a/terraform/node_resource_apply_instance.go
+++ b/terraform/node_resource_apply_instance.go
@@ -10,6 +10,7 @@ import (
 	"github.com/hashicorp/terraform/plans"
 	"github.com/hashicorp/terraform/providers"
 	"github.com/hashicorp/terraform/states"
+	"github.com/hashicorp/terraform/tfdiags"
 )
 
 // NodeApplyableResourceInstance represents a resource instance that is
@@ -106,6 +107,27 @@ func (n *NodeApplyableResourceInstance) EvalTree() EvalNode {
 
 	// Determine the dependencies for the state.
 	stateDeps := n.StateReferences()
+
+	if n.Config == nil {
+		// This should not be possible, but we've got here in at least one
+		// case as discussed in the following issue:
+		//    https://github.com/hashicorp/terraform/issues/21258
+		// To avoid an outright crash here, we'll instead return an explicit
+		// error.
+		var diags tfdiags.Diagnostics
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Resource node has no configuration attached",
+			fmt.Sprintf(
+				"The graph node for %s has no configuration attached to it. This suggests a bug in Terraform's apply graph builder; please report it!",
+				addr,
+			),
+		))
+		err := diags.Err()
+		return &EvalReturnError{
+			Error: &err,
+		}
+	}
 
 	// Eval info is different depending on what kind of resource this is
 	switch n.Config.Mode {

